### ISSUE TITLE
How are experiments created through the GUI structured?

### ISSUE THREAD

**Issue by jdpigeon** [jspsych/jsPsych-Redux-GUI#227]
Hello there,

I'm working on an app, [BrainWaves](github.com/makebrainwaves/BrainWaves), that will make heavy use of jspsych in order to allow students to perform and design their own neuroscience experiments.

I'm trying right now to design a system for reading timelines in my app so that they can be displayed through a [jspsych-react](github.com/makebrainwaves/jspsych-react) component. I don't want to reinvent the wheel, however, and I'd like my app to be compatible with the other tools people will be using to create jsPsych experiments such as the Redux GUI.

One source of inspiration is the [Experiment Factory experiments](https://github.com/exprxiv). However, I find the design of those experiments very rigid and imperative. I would like to move to a design that's more flexible and allows users to easily modify timeline parameters without code.

Because I'm using the same technologies as this GUI (React + Redux), I'd love to see what data structures are used when saving user-created timelines. Is there an example of that? I wasn't able to find anything briefly poking through the source code and playing with the app itself

**Comment by jdpigeon** [comment]
Awesome, That's a very nice way to do it! I assume that tree graph structure is preserved when saving the experiment file to disk?

And yes, I hope to make use of this GUI as much as possible. How that will happen, I'm currently unsure; I definitely want to slim down the experiment creation experience to only contain what's relevant for our specific EEG-based use case.

I will stay in touch! Cheers!

**Comment by GavinQ1** [comment]
Yes, the tree graph structure is preserved since the order of elements in array is always preserved. But there is one downside for this kind of structure. It introduces many complications when adding, moving (from one position to another position in the tree) and deleting a node with children. It won't be a problem if this tree graph is traditional (nested). However, with the flattened data structure, implementation is cumbersome.

**Comment by jdpigeon** [comment]
Hey Gavin,

Here's an example of the data structure I came up with for storing jsPsych experiments in Redux. I think it's a little different from what you did so I thought I'd share. Note that I've included jsPsych variables.

Also, @teonbrooks

```javascript
const experimentState = {
  mainTimeline: ["welcome", "faceHouseTimeline", "end"], // array of trial and timeline ids
  trials: {
    welcome: {
      type: "callback_html_display",
      id: "welcome",
      stimulus: "Welcome to the experiment. Press any key to begin.",
      post_trial_gap: 1000,
      on_load: () => console.log("start")
    },
    end: {
      id: "end",
      type: "callback_html_display",
      stimulus: "Thanks for participating",
      post_trial_gap: 500,
      on_load: console.log("stop")
    }
  },
  timelines: {
    faceHouseTimeline: {
      id: "faceHouseTimeline",
      timeline: [
        {
          type: "callback_image_display",
          stimulus: "./assets/face_house/fixation.jpg",
          trial_duration: () => 300 + Math.random() * 200
        },
        {
          stimulus: jsPsych.timelineVariable("stimulusVar"),
          on_load: jsPsych.timelineVariable("callbackVar"),
          type: "callback_image_display",
          choices: ["f", "j"],
          trial_duration: 300
        }
      ],
      sample: {
        type: "with-replacement",
        size: 170
      },
      timeline_variables: [
        // list of image paths and callback fns
        {
          stimulusVar: "./assets/face_house/faces/Andrew.jpg",
          callbackVar: () => console.log("face")
        },
        {
          stimulusVar: "./assets/face_house/houses/House_1.jpg",
          callbackVar: () => console.log("house")
        }
      ]
    }
  }
};

```

And to parse this into a legitimate jsPsych timeline that the library can accept is super easy:
```javascript
// ./jspsych/functions.js
// Converts a normalized timeline template into a classic jsPsych timeline array
export const parseTimeline = (mainTimeline, trials, timelines) => {
  const jsPsychObject = { ...trials, ...timelines };
  return mainTimeline.map(id => jsPsychObject[id]);
};
```

**Comment by GavinQ1** [comment]
hi Dano, that looks good! However, I'd still suggest use id to refer to trial/timeline objects (consider it like pointer in C/C++ or reference in Java). For example, when you want to nest another timeline (who also has nested timelines. let's be extreme here ;) ) and then modify one trial that is nested in some deep nested timeline, it will be hard to identify the wanted trial and modify it while complying to the idea of not mutating state in redux  (unless making a deep copy of the whole thing every time).

**Comment by GavinQ1** [comment]
As for pares timeline to jsPsych format, a simple dfs would work.

**Comment by jdpigeon** [comment]
I agree that everything should be referenced by id and I've done that here except for the trials inside the secondary timeline. I was hesitant to break those out on their own because I'm not entirely sure how the jsPsych timeline variables work.

I would argue that putting trials and timelines into objects instead of having them at the root level in the store is more Redux-y. I believe it makes type checking and mapping state to React props easier to understand

**Comment by GavinQ1** [comment]
Ah, now i see what are `trails` and `timelines` for. As for type checking, I kept a ```type``` property in each node. In fact, there is a inheritance relationship here (although DynamoDB keeps me from declaring a actual ES6 class in real implementation). TimelineNode class and TrialNode class both extend GraphNode class. A graphNode instance is responsible for holding generic information (e.g. id, children) and extra GUI information. I did that because merely holding jsPsych related information is not enough for our desired features. Otherwise, your way would make mapping much easier. :)
